descartes_moveit/src/seed_search.cpp

```diff
@@ -0,0 +1,455 @@
+{
+  for (std::size_t i = 0; i < lhs.size(); ++i)
+  {
+    if ( lhs[i] != rhs[i] )
+    {
+      return lhs[i] < rhs[i];
+    }
+  }
+  // equal
+  return false;
+}
+
+// Simple container for holding info necessary for converting from handles to joint configs
+class ConfigCreator
+{
+public:
+  /**
+   * @brief ConfigCreator constructor
+   * @param initial_joint_pos The values for all joints that are not being permuted
+   * @param bij The bijection vector where the first values in the bijection are the
+   *            joints being permuted
+   * @param min_angle The lower limit of the joint angles
+   * @param step_angle Joint angle discretization
+   */
+  ConfigCreator(const std::vector<double> initial_joint_pos,
+                const BijectionVec& bij, double min_angle, double step_angle)
+    : initial_joint_positions_(initial_joint_pos)
+    , bij_(bij)
+    , min_angle_(min_angle)
+    , step_angle_(step_angle)
+  {}
+
+  seed::JointConfig toJointConfig(const ComboHandle& handle) const
+  {
+    seed::JointConfig result;
+    result.resize(bij_.size());
+
+    std::size_t i;
+    for (i = 0; i < handle.size(); i++)
+    {
+      result[bij_[i]] = min_angle_ + handle[i] * step_angle_;
+    }
+
+    for (; i < bij_.size(); ++i)
+    {
+      result[bij_[i]] = initial_joint_positions_[bij_[i]];
+    }
+
+    return result;
+  }
+
+private:
+  std::vector<double> initial_joint_positions_;
+  BijectionVec bij_;
+  double min_angle_;
+  double step_angle_;
+};
+
+/**
+ * @brief Forward kinematics helper
+ */
+bool doFK(moveit::core::RobotState& state,
+          const moveit::core::JointModelGroup* group,
+          const std::string& tool,
+          const seed::JointConfig& joint_pose,
+          Eigen::Affine3d& result)
+{
+  state.setJointGroupPositions(group, joint_pose);
+  if (!state.knowsFrameTransform(tool))
+  {
+    ROS_WARN("No transform to this tool frame");
+    return false;
+  }
+
+  if (!state.satisfiesBounds())
+  {
+    ROS_WARN("Joint angles do not satisfy robot bounds");
+    return false;
+  }
+
+  result = state.getFrameTransform(tool);
+  return true;
+}
+
+/**
+ * @brief Inverse kinematics helper. Returns the solution in the result parameter.
+ *        May return false if IK failed.
+ */
+bool doIK(moveit::core::RobotState& state,
+          const moveit::core::JointModelGroup* group,
+          const std::string& group_name,
+          const std::string& tool,
+          const Eigen::Affine3d& pose,
+          const seed::JointConfig& seed,
+          seed::JointConfig& result)
+{
+  state.setJointGroupPositions(group_name, seed);
+  if (!state.setFromIK(group, pose, tool, 1, 0.01))
+  {
+    return false;
+  }
+  state.copyJointGroupPositions(group, result);
+  return true;
+}
+
+/**
+ * @brief Returns true if the determinant of the jacobian is not zero.
+ *        Returns false if it is close to zero (and robot is near singularity)
+ */
+bool checkSingularity(moveit::core::RobotState& state,
```

Comment:
Just a suggestion...it's more clear to use an `is*` name.  So for here, is should be `isNonsingular`.